crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3431,6 +3452,139 @@ impl<'db> TypeInferenceBuilder<'db> {
+                    Type::Never
+                    | Type::Unknown
+                    | Type::Unbound
+                    | Type::Todo
+                    | Type::FunctionLiteral(_)
+                    | Type::ModuleLiteral(_)
+                    | Type::ClassLiteral(_)
+                    | Type::Union(_)
+                    | Type::Intersection(_)
+                    | Type::Any => {
+                        self.add_diagnostic(
+                                    parameters.into(),
+                                    "invalid-literal-parameter",
+                                    format_args!(
+                                        "Type arguments for `Literal` must be None, a literal value (int, bool, str, or bytes), or an enum value",
+                                    ),
+                                );
+                        Type::Unknown
+                    }
+                    Type::Tuple(tuple) => {
+                        let elts = tuple.elements(self.db);
+                        Type::Union(UnionType::new(self.db, elts))
+                    }
+                    ty => ty,
+                }
+            }
+        }
+    }
+
+    fn infer_literal_parameter_type(&mut self, parameters: &ast::Expr) -> Type<'db> {
+        // If the slice itself is another subscript and it is another Literal then flatten
+        // the values
+        match parameters {
+            ruff_python_ast::Expr::Subscript(inner_literal_subscript) => {
+                let inner_subscript_value = self.infer_expression(&inner_literal_subscript.value);
+                if let Type::Instance(inner_instance) = inner_subscript_value {
+                    if inner_instance.is_known(KnownInstance::Literal) {
+                        self.infer_parameterized_known_instance_type_expression(
+                            KnownInstance::Literal,
+                            &inner_literal_subscript.slice,
+                        )
+                    } else {
+                        Type::Unknown
+                    }
+                } else {
+                    self.add_diagnostic(
+                                        parameters.into(),
+                                        "invalid-literal-parameter",
+                                        format_args!(
+                                            "Type arguments for `Literal` must be None, a literal value (int, bool, str, or bytes), or an enum value",
+                                        ),
+                                    );
+                    Type::Unknown
+                }
+            }
+            ruff_python_ast::Expr::Tuple(t) => {
+                let mut elts = vec![];
+                for elm in &t.elts {
+                    elts.push(self.infer_literal_parameter_type(elm));
+                }
+                Type::Tuple(TupleType::new(self.db, elts.into_boxed_slice()))
+            }
+            _ => self.infer_expression(parameters),
```

Comment:
Rather than matching out in `infer_parameterized_known_instance_type_expression` on AST forms known not to be valid, I think we should explicitly match here on each AST form known to be valid, and directly return the right type, without relying on `self.infer_expression`.

Reply:
Got it. Do you mean to not rely only on `infer_expression` so I call each ast type infer method separately? for example for boolean call:
```
ruff_python_ast::Expr::BooleanLiteral(literal) => {
                self.infer_boolean_literal_expression(literal)
            }
```

Or do you mean to completely not rely on those functions and return the type directly? I did use the functions to avoid repeating them could something go wrong this way?